### PR TITLE
Fix getauxval for sanitizer builds

### PR DESCRIPTION
**Pull request by amosbird**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix segfault when certain sanitizer is used. cc @azat 

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: base/glibc-compatibility/musl/getauxval.c [hunk 1]
```diff
@@ -39,7 +41,9 @@ ssize_t __retry_read(int fd, void * buf, size_t count)
 {
     for (;;)
     {
-        ssize_t ret = read(fd, buf, count);
+        // We cannot use the read syscall as it will be intercept by sanitizers, which aren't
+        // initialized yet. Emit syscall directly.
```
Comment (azat) [inline base/glibc-compatibility/musl/getauxval.c:45]: Shouldn't the same be done for `close` ?
BTW does it fixes your issue (on non x86_64)?

Comment (amosbird) [reply base/glibc-compatibility/musl/getauxval.c:45]: > Shouldn't the same be done for `close`?

Sanitizers likely intercept buffer related syscalls. `open` and `close` don't fit in this category.

> BTW does it fixes your issue (on non x86_64)?

Yes.

**Review by tavplubix — APPROVED** [review]

**Comment by tavplubix** [comment]
Stateless tests (msan) [3/3] failure is suspicious:
```
==648==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x5edfe5f8 in __auxv_init_procfs build_docker/../base/glibc-compatibility/musl/getauxval.c:95:5
    #1 0x5edfdfd8 in getauxval build_docker/../base/glibc-compatibility/musl/getauxval.c:197:12
    #2 0x5ee0ab26 in __vdsosym build_docker/../base/glibc-compatibility/musl/vdso.c:46:23
    #3 0x5ee09c61 in getcpu_init build_docker/../base/glibc-compatibility/musl/sched_getcpu.c:21:12
    #4 0x5ee09a3c in sched_getcpu build_docker/../base/glibc-compatibility/musl/sched_getcpu.c:39:7
    #5 0x5752c1d6 in rocksdb::port::PhysicalCoreID() build_docker/../contrib/rocksdb/port/port_posix.cc:164:15
    #6 0x56ef476b in rocksdb::CoreLocalArray<rocksdb::StatisticsImpl::StatisticsData>::AccessElementAndIndex() const build_docker/../contrib/rocksdb/util/core_local.h:66:15
    #7 0x56ef476b in rocksdb::CoreLocalArray<rocksdb::StatisticsImpl::StatisticsData>::Access() const build_docker/../contrib/rocksdb/util/core_local.h:61:10
    #8 0x56ef476b in rocksdb::StatisticsImpl::recordInHistogram(unsigned int, unsigned long) build_docker/../contrib/rocksdb/monitoring/statistics.cc:370:19
    #9 0x56ef78bf in rocksdb::Statistics::reportTimeToHistogram(unsigned int, unsigned long) build_docker/../contrib/rocksdb/include/rocksdb/statistics.h:561:5
    #10 0x56704047 in rocksdb::StopWatch::~StopWatch() build_docker/../contrib/rocksdb/util/stop_watch.h:47:20
    #11 0x56e6e43f in rocksdb::SyncManifest(rocksdb::ImmutableDBOptions const*, rocksdb::WritableFileWriter*) build_docker/../contrib/rocksdb/file/filename.cc:429:1
    #12 0x564c0af7 in rocksdb::DBImpl::NewDB(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>*) build_docker/../contrib/rocksdb/db/db_impl/db_impl_open.cc:310:11
    #13 0x564c813c in rocksdb::DBImpl::Recover(std::__1::vector<rocksdb::ColumnFamilyDescriptor, std::__1::allocator<rocksdb::ColumnFamilyDescriptor>> const&, bool, bool, bool, unsigned long*) build_docker/../contrib/rocksdb/db/db_impl/db_impl_open.cc:426:13
    #14 0x564f00aa in rocksdb::DBImpl::Open(rocksdb::DBOptions const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::vector<rocksdb::ColumnFamilyDescriptor, std::__1::allocator<rocksdb::ColumnFamilyDescriptor>> const&, std::__1::vector<rocksdb::ColumnFamilyHandle*, std::__1::allocator<rocksdb::ColumnFamilyHandle*>>*, rocksdb::DB**, bool, bool) build_docker/../contrib/rocksdb/db/db_impl/db_impl_open.cc:1615:13
    #15 0x564ebf1a in rocksdb::DB::Open(rocksdb::DBOptions const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::vector<rocksdb::ColumnFamilyDescriptor, std::__1::allocator<rocksdb::ColumnFamilyDescriptor>> const&, std::__1::vector<rocksdb::ColumnFamilyHandle*, std::__1::allocator<rocksdb::ColumnFamilyHandle*>>*, rocksdb::DB**) build_docker/../contrib/rocksdb/db/db_impl/db_impl_open.cc:1505:10
    #16 0x564ebf1a in rocksdb::DB::Open(rocksdb::Options const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, rocksdb::DB**) build_docker/../contrib/rocksdb/db/db_impl/db_impl_open.cc:1483:14
    #17 0x3df137b2 in DB::StorageEmbeddedRocksDB::initDB() build_docker/../src/Storages/RocksDB/StorageEmbeddedRocksDB.cpp:388:22
    #18 0x3df0ec3f in DB::StorageEmbeddedRocksDB::StorageEmbeddedRocksDB(DB::StorageID const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, DB::StorageInMemoryMetadata const&, bool, std::__1::shared_ptr<DB::Context const>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, bool) build_docker/../src/Storages/RocksDB/StorageEmbeddedRocksDB.cpp:194:5
    #19 0x3df35135 in DB::StorageEmbeddedRocksDB* std::__1::construct_at[abi:v15003]<DB::StorageEmbeddedRocksDB, DB::StorageID const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, DB::StorageInMemoryMetadata&, bool const&, std::__1::shared_ptr<DB::Context>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&, int&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, bool&, DB::StorageEmbeddedRocksDB*>(DB::StorageEmbeddedRocksDB*, DB::StorageID const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, DB::StorageInMemoryMetadata&, bool const&, std::__1::shared_ptr<DB::Context>&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&, int&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&&, bool&) build_docker/../contrib/libcxx/include/__memory/construct_at.h:35:48
    #20 0x3df35135 in void std::__1::allocator_traits<std::__1::allocator<DB::StorageEmbeddedRocksDB>>::construct[abi:v15003]<DB::StorageEmbeddedRocksDB, DB::StorageID const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, DB::StorageInMemoryMetadata&, bool const&, std::__1::shared_ptr<DB::Context>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&, int&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, bool&, void, void>(std::__1::allocator<DB::StorageEmbeddedRocksDB>&, DB::StorageEmbeddedRocksDB*, DB::StorageID const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, DB::StorageInMemoryMetadata&, bool const&, std::__1::shared_ptr<DB::Context>&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&, int&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&&, bool&) build_docker/../contrib/libcxx/include/__memory/allocator_traits.h:298:9
    #21 0x3df34b19 in std::__1::__shared_ptr_emplace<DB::StorageEmbeddedRocksDB, std::__1::allocator<DB::StorageEmbeddedRocksDB>>::__shared_ptr_emplace[abi:v15003]<DB::StorageID const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, DB::StorageInMemoryMetadata&, bool const&, std::__1::shared_ptr<DB::Context>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&, int&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, bool&>(std::__1::allocator<DB::StorageEmbeddedRocksDB>, DB::StorageID const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, DB::StorageInMemoryMetadata&, bool const&, std::__1::shared_ptr<DB::Context>&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&, int&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&&, bool&) build_docker/../contrib/libcxx/include/__memory/shared_ptr.h:292:9
    #22 0x3df34b19 in std::__1::shared_ptr<DB::StorageEmbeddedRocksDB> std::__1::allocate_shared[abi:v15003]<DB::StorageEmbeddedRocksDB, std::__1::allocator<DB::StorageEmbeddedRocksDB>, DB::StorageID const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, DB::StorageInMemoryMetadata&, bool const&, std::__1::shared_ptr<DB::Context>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&, int&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, bool&, void>(std::__1::allocator<DB::StorageEmbeddedRocksDB> const&, DB::StorageID const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, DB::StorageInMemoryMetadata&, bool const&, std::__1::shared_ptr<DB::Context>&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&, int&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&&, bool&) build_docker/../contrib/libcxx/include/__memory/shared_ptr.h:953:55
    #23 0x3df23bcb in std::__1::shared_ptr<DB::StorageEmbeddedRocksDB> std::__1::make_shared[abi:v15003]<DB::StorageEmbeddedRocksDB, DB::StorageID const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, DB::StorageInMemoryMetadata&, bool const&, std::__1::shared_ptr<DB::Context>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&, int&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, bool&, void>(DB::StorageID const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, DB::StorageInMemoryMetadata&, bool const&, std::__1::shared_ptr<DB::Context>&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&, int&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&&, bool&) build_docker/../contrib/libcxx/include/__memory/shared_ptr.h:962:12
    #24 0x3df23bcb in DB::create(DB::StorageFactory::Arguments const&) build_docker/../src/Storages/RocksDB/StorageEmbeddedRocksDB.cpp:489:12
    #25 0x3df486e3 in decltype(std::declval<std::__1::shared_ptr<DB::IStorage> (*&)(DB::StorageFactory::Arguments const&)>()(std::declval<DB::StorageFactory::Arguments const&>())) std::__1::__invoke[abi:v15003]<std::__1::shared_ptr<DB::IStorage> (*&)(DB::StorageFactory::Arguments const&), DB::StorageFactory::Arguments const&>(std::__1::shared_ptr<DB::IStorage> (*&)(DB::StorageFactory::Arguments const&), DB::StorageFactory::Arguments const&) build_docker/../contrib/libcxx/include/__functional/invoke.h:394:23
    #26 0x3df486e3 in std::__1::shared_ptr<DB::IStorage> std::__1::__invoke_void_return_wrapper<std::__1::shared_ptr<DB::IStorage>, false>::__call<std::__1::shared_ptr<DB::IStorage> (*&)(DB::StorageFactory::Arguments const&), DB::StorageFactory::Arguments const&>(std::__1::shared_ptr<DB::IStorage> (*&)(DB::StorageFactory::Arguments const&), DB::StorageFactory::Arguments const&) build_docker/../contrib/libcxx/include/__functional/invoke.h:470:16
    #27 0x3df486e3 in std::__1::__function::__default_alloc_func<std::__1::shared_ptr<DB::IStorage> (*)(DB::StorageFactory::Arguments const&), std::__1::shared_ptr<DB::IStorage> (DB::StorageFactory::Arguments const&)>::operator()[abi:v15003](DB::StorageFactory::Arguments const&) build_docker/../contrib/libcxx/include/__functional/function.h:235:12
    #28 0x3df486e3 in std::__1::shared_ptr<DB::IStorage> std::__1::__function::__policy_invoker<std::__1::shared_ptr<DB::IStorage> (DB::StorageFactory::Arguments const&)>::__call_impl<std::__1::__function::__default_alloc_func<std::__1::shared_ptr<DB::IStorage> (*)(DB::StorageFactory::Arguments const&), std::__1::shared_ptr<DB::IStorage> (DB::StorageFactory::Arguments const&)>>(std::__1::__function::__policy_storage const*, DB::StorageFactory::Arguments const&) build_docker/../contrib/libcxx/include/__functional/function.h:716:16
    #29 0x4544379a in std::__1::__function::__policy_func<std::__1::shared_ptr<DB::IStorage> (DB::StorageFactory::Arguments const&)>::operator()[abi:v15003](DB::StorageFactory::Arguments const&) const build_docker/../contrib/libcxx/include/__functional/function.h:848:16
    #30 0x4544379a in std::__1::function<std::__1::shared_ptr<DB::IStorage> (DB::StorageFactory::Arguments const&)>::operator()(DB::StorageFactory::Arguments const&) const build_docker/../contrib/libcxx/include/__functional/function.h:1197:12
    #31 0x4544379a in DB::StorageFactory::get(DB::ASTCreateQuery const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context>, std::__1::shared_ptr<DB::Context>, DB::ColumnsDescription const&, DB::ConstraintsDescription const&, bool) const build_docker/../src/Storages/StorageFactory.cpp:229:16
    #32 0x427406e5 in DB::InterpreterCreateQuery::doCreateTable(DB::ASTCreateQuery&, DB::InterpreterCreateQuery::TableProperties const&, std::__1::unique_ptr<DB::DDLGuard, std::__1::default_delete<DB::DDLGuard>>&) build_docker/../src/Interpreters/InterpreterCreateQuery.cpp:1356:42
    #33 0x4272eae9 in DB::InterpreterCreateQuery::createTable(DB::ASTCreateQuery&) build_docker/../src/Interpreters/InterpreterCreateQuery.cpp:1220:20
    #34 0x4274e417 in DB::InterpreterCreateQuery::execute() build_docker/../src/Interpreters/InterpreterCreateQuery.cpp:1630:16
    #35 0x43d76f57 in DB::executeQueryImpl(char const*, char const*, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) build_docker/../src/Interpreters/executeQuery.cpp:709:36
    #36 0x43d6acbd in DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum) build_docker/../src/Interpreters/executeQuery.cpp:1106:30
    #37 0x47466155 in DB::TCPHandler::runImpl() build_docker/../src/Server/TCPHandler.cpp:374:24
    #38 0x474a725d in DB::TCPHandler::run() build_docker/../src/Server/TCPHandler.cpp:1902:9
    #39 0x531c4e7d in Poco::Net::TCPServerConnection::start() build_docker/../contrib/poco/Net/src/TCPServerConnection.cpp:43:3
    #40 0x531c620e in Poco::Net::TCPServerDispatcher::run() build_docker/../contrib/poco/Net/src/TCPServerDispatcher.cpp:115:20
    #41 0x5392726b in Poco::PooledThread::run() build_docker/../contrib/poco/Foundation/src/ThreadPool.cpp:199:14
    #42 0x53922801 in Poco::(anonymous namespace)::RunnableHolder::run() build_docker/../contrib/poco/Foundation/src/Thread.cpp:55:11
    #43 0x5391e5c8 in Poco::ThreadImpl::runnableEntry(void*) build_docker/../contrib/poco/Foundation/src/Thread_POSIX.cpp:345:27
    #44 0x7fbb52ecf608 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x8608) (BuildId: 7b4536f41cdaa5888408e82d0836e33dcf436466)
    #45 0x7fbb52df4132 in __clone (/lib/x86_64-linux-gnu/libc.so.6+0x11f132) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)

  Uninitialized value was created by an allocation of 'aux' in the stack frame of function '__auxv_init_procfs'
    #0 0x5edfdfe0 in __auxv_init_procfs build_docker/../base/glibc-compatibility/musl/getauxval.c:75

SUMMARY: MemorySanitizer: use-of-uninitialized-value build_docker/../base/glibc-compatibility/musl/getauxval.c:95:5 in __auxv_init_procfs
```

**Review by tavplubix — APPROVED** [review]

File: base/glibc-compatibility/musl/getauxval.c [hunk 2]
```diff
@@ -90,6 +94,11 @@ static unsigned long NO_SANITIZE_THREAD __auxv_init_procfs(unsigned long type)
     _Static_assert(sizeof(aux) < 4096, "Unexpected sizeof(aux)");
     while (__retry_read(fd, &aux, sizeof(aux)) == sizeof(aux))
     {
+#if defined(__has_feature)
+#if __has_feature(memory_sanitizer)
+        __msan_unpoison(&aux, sizeof(aux));
```
Comment (Algunenano) [inline base/glibc-compatibility/musl/getauxval.c:99]: This has introduced a new warning:

```
[30/10859] Building C object base/glibc-compatibility/CMakeFiles/glibc-compatibility.dir/musl/getauxval.c.o
/mnt/ch/ClickHouse/base/glibc-compatibility/musl/getauxval.c:99:9: warning: implicit declaration of function '__msan_unpoison' is invalid in C99 [-Wimplicit-function-declaration]
        __msan_unpoison(&aux, sizeof(aux));
```

Comment (tavplubix) [reply base/glibc-compatibility/musl/getauxval.c:99]: But all builds were ok...

Comment (tavplubix) [reply base/glibc-compatibility/musl/getauxval.c:99]: > I don't know if this is the desired behaviour for those builds though.

I think it's not, we should either suppress/disable unneeded warnings or turn them into errors

Comment (Algunenano) [reply base/glibc-compatibility/musl/getauxval.c:99]: I think I was wrong and what's likely happening is that base/ is not covered by -Werror. For the moment I've created a PR to address the warning (https://github.com/ClickHouse/ClickHouse/pull/43065)

Comment (alexey-milovidov) [reply base/glibc-compatibility/musl/getauxval.c:99]: @Algunenano all ClickHouse code should be covered by -Werror.
Only third-party libraries are exceptions.

Comment (Algunenano) [reply base/glibc-compatibility/musl/getauxval.c:99]: PR for that here https://github.com/ClickHouse/ClickHouse/pull/43082. It's already detecting issues

**Pull request merged**